### PR TITLE
Fix roundchange behavior when the number of nodes are less than 4.

### PR DESCRIPTION
**Pull request by andybclee**
## Proposed changes

- This PR is to solve an issue of block generation when F() is 0 (meaning that the number of nodes are less than 4)
- When a roundchange happened, the required number of messages are calculated differently based on the number of nodes

## Types of changes

Please put an x in the boxes related to your change.

- [x] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by aidan-kwon** [comment]
@andybclee It seems that `handleRoundChange` can be triggered by the protocol message from other nodes. When the validator set size is 1, how does it change the round without any received message?

**Comment by andybclee** [comment]
> @andybclee It seems that `handleRoundChange` can be triggered by the protocol message from other nodes. When the validator set size is 1, how does it change the round without any received message?

When starting a new round, each node set a timeout timer using `time.AfterFunc`. In this way, each node set its own round based on a timer.

File: consensus/istanbul/core/roundchange.go [hunk 1]
```diff
@@ -101,20 +101,35 @@ func (c *core) handleRoundChange(msg *message, src istanbul.Validator) error {
...
+	var numCatchUp, numStartNewRound int
+	if c.valSet.Size() < 4 {
+		n := int(c.valSet.Size())
+		// N ROUND CHANGE messages can start new round.
+		numStartNewRound = n
+		// N - 1 ROUND CHANGE messages can catch up the round.
+		numCatchUp = n - 1
+	} else {
+		f := int(c.valSet.F())
+		// 2*f + 1 ROUND CHANGE messages can start new round.
```
Comment (KimKyungup) [inline consensus/istanbul/core/roundchange.go]:
```suggestion
		// 2*F + 1 ROUND CHANGE messages can start new round.
```

File: consensus/istanbul/core/roundchange.go [hunk 2]
```diff
@@ -101,20 +101,35 @@ func (c *core) handleRoundChange(msg *message, src istanbul.Validator) error {
...
+		numCatchUp = n - 1
+	} else {
+		f := int(c.valSet.F())
+		// 2*F + 1 ROUND CHANGE messages can start new round.
+		numStartNewRound = 2*f + 1
+		// F + 1 ROUND CHANGE messages can start catch up the round.
+		numCatchUp = f + 1
+	}
+
+	if num == numStartNewRound && (c.waitingForRoundChange || cv.Round.Cmp(roundView.Round) < 0) {
```
Comment (jeonghonah) [inline consensus/istanbul/core/roundchange.go]: In case of single node configuration, which conditions cause to `false` in the if-condition?

**Comment by aidan-kwon** [comment]
@andybclee Thank you for the explanation.
The code looks good to me, but we may choose another approach. 
How do you think about making `valSet.F()` to return 0 when `valSet.Size()` is less than 4.

File: consensus/istanbul/core/roundchange.go (hunk 2 above)
Comment (andybclee) [reply consensus/istanbul/core/roundchange.go]: @jeonghonah In this line, all conditions will be true in the single node configuration. So the node can start the new round with a new proposal.

**Comment by andybclee** [comment]
> @andybclee Thank you for the explanation.
> The code looks good to me, but we may choose another approach.
> How do you think about making `valSet.F()` to return 0 when `valSet.Size()` is less than 4.

@aidan-kwon As you can see below, F() returns 0 when the size is smaller than 4.
```go
func (valSet *weightedCouncil) F() int {
	if valSet.Size() > valSet.subSize {
		return int(math.Ceil(float64(valSet.subSize)/3)) - 1
	} else {
		return int(math.Ceil(float64(valSet.Size())/3)) - 1
	}
}
```

**Comment by aidan-kwon** [comment]
@andybclee Oh... I misunderstood about it. Thank you!

File: consensus/istanbul/core/roundchange.go [hunk 3]
```diff
@@ -101,20 +101,35 @@ func (c *core) handleRoundChange(msg *message, src istanbul.Validator) error {
...
+		// We've received enough ROUND CHANGE messages, start a new round immediately.
+		logger.Warn("[RC] Received 2f+1 Round Change Messages. Starting new round")
+		c.startNewRound(roundView.Round)
+		return nil
+	} else if c.waitingForRoundChange && num == numCatchUp {
```
Comment (jeonghonah) [inline consensus/istanbul/core/roundchange.go]: Is it not necessary to check `cv.Round.Cmp(roundView.Round) < 0` condition here?

**Pull request merged**